File: data_mcp/gpt-5/scala/rep5/src/Main.scala
//> using scala "3.3.1"
//> using dep "org.http4s::http4s-ember-server:0.23.34"
//> using dep "org.http4s::http4s-dsl:0.23.34"
//> using dep "org.http4s::http4s-circe:0.23.34"
//> using dep "io.circe::circe-core:0.14.15"
//> using dep "io.circe::circe-generic:0.14.15"
//> using dep "io.circe::circe-parser:0.14.15"

import cats.effect._
import cats.syntax.all._
import org.http4s._
import org.http4s.dsl.io._
import org.http4s.dsl.Http4sDsl
import org.http4s.ember.server._
import org.http4s.implicits._
import org.http4s.circe._
import org.http4s.circe.CirceEntityDecoder._
import org.http4s.circe.CirceEntityEncoder._
import io.circe._
import io.circe.syntax._
import io.circe.generic.semiauto._
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import scala.jdk.CollectionConverters._
import java.time.Instant
import java.time.temporal.ChronoUnit
import com.comcast.ip4s._

object Main extends IOApp {

  final case class User(id: Int, username: String, password: String)
  object User {
    implicit val userEncoder: Encoder[User] = new Encoder[User] {
      def apply(u: User): Json = Json.obj(
        "id" -> Json.fromInt(u.id),
        "username" -> Json.fromString(u.username)
      )
    }
  }

  final case class Todo(
      id: Int,
      title: String,
      description: String,
      completed: Boolean,
      created_at: String,
      updated_at: String,
      ownerId: Int
  )
  object Todo {
    implicit val todoEncoder: Encoder[Todo] = new Encoder[Todo] {
      def apply(t: Todo): Json = Json.obj(
        "id" -> Json.fromInt(t.id),
        "title" -> Json.fromString(t.title),
        "description" -> Json.fromString(t.description),
        "completed" -> Json.fromBoolean(t.completed),
        "created_at" -> Json.fromString(t.created_at),
        "updated_at" -> Json.fromString(t.updated_at)
      )
    }
  }

  // Request DTOs
  final case class RegisterReq(username: String, password: String)
  object RegisterReq { implicit val dec: Decoder[RegisterReq] = deriveDecoder }

  final case class LoginReq(username: String, password: String)
  object LoginReq { implicit val dec: Decoder[LoginReq] = deriveDecoder }

  final case class PasswordChangeReq(old_password: String, new_password: String)
  object PasswordChangeReq { implicit val dec: Decoder[PasswordChangeReq] = deriveDecoder }

  final case class CreateTodoReq(title: String, description: Option[String])
  object CreateTodoReq { implicit val dec: Decoder[CreateTodoReq] = deriveDecoder }

  final case class UpdateTodoReq(title: Option[String], description: Option[String], completed: Option[Boolean])
  object UpdateTodoReq { implicit val dec: Decoder[UpdateTodoReq] = deriveDecoder }

  // In-memory stores
  private val userIdGen = new AtomicInteger(0)
  private val todoIdGen = new AtomicInteger(0)
  private val usersById = new ConcurrentHashMap[Int, User]()
  private val usersByUsername = new ConcurrentHashMap[String, User]()
  private val sessions = new ConcurrentHashMap[String, Int]() // token -> userId
  private val todosById = new ConcurrentHashMap[Int, Todo]()
  private val usersLock = new Object()

  private val cookieName = "session_id"

  private def nowTs(): String = Instant.now().truncatedTo(ChronoUnit.SECONDS).toString

  private def jsonError(status: Status, msg: String): IO[Response[IO]] = {
    val body = Json.obj("error" -> Json.fromString(msg))
    IO.pure(Response[IO](status).withEntity(body))
  }

  private def readJson[A: Decoder](req: Request[IO]): IO[Either[Response[IO], A]] =
    req.attemptAs[Json].value.flatMap {
      case Left(_) => jsonError(Status.BadRequest, "Invalid request").map(Left(_))
      case Right(json) =>
        json.as[A] match {
          case Left(_)  => jsonError(Status.BadRequest, "Invalid request").map(Left(_))
          case Right(a) => IO.pure(Right(a))
        }
    }

  private def withAuth(req: Request[IO])(f: User => IO[Response[IO]]): IO[Response[IO]] = {
    val maybeToken = req.cookies.find(_.name == cookieName).map(_.content)
    maybeToken match {
      case None => jsonError(Status.Unauthorized, "Authentication required")
      case Some(token) =>
        val uid = sessions.get(token)
        if (uid == 0 && !sessions.containsKey(token)) then
          jsonError(Status.Unauthorized, "Authentication required")
        else {
          val user = usersById.get(uid)
          if (user == null) jsonError(Status.Unauthorized, "Authentication required")
          else f(user)
        }
    }
  }

  private def register(username: String, password: String): Either[Response[IO], User] = {
    val validUser = username.matches("^[a-zA-Z0-9_]{3,50}$")
    if (!validUser) return Left(Response[IO](Status.BadRequest).withEntity(Json.obj("error" -> Json.fromString("Invalid username"))))
    if (password.length < 8) return Left(Response[IO](Status.BadRequest).withEntity(Json.obj("error" -> Json.fromString("Password too short"))))

    usersLock.synchronized {
      if (usersByUsername.containsKey(username))
        Left(Response[IO](Status.Conflict).withEntity(Json.obj("error" -> Json.fromString("Username already exists"))))
      else {
        val id = userIdGen.incrementAndGet()
        val user = User(id, username, password)
        usersById.put(id, user)
        usersByUsername.put(username, user)
        Right(user)
      }
    }
  }

  private def createSession(userId: Int): String = {
    val token = UUID.randomUUID().toString.replaceAll("-", "")
    sessions.put(token, userId)
    token
  }

  private def invalidateSession(token: String): Unit = {
    sessions.remove(token)
  }

  private def parseTodoId(s: String): Option[Int] =
    try Some(s.toInt) catch { case _: NumberFormatException => None }

  private def routes: HttpRoutes[IO] = {
    val dsl = new Http4sDsl[IO] {}
    import dsl._

    HttpRoutes.of[IO] {
      // Register
      case req @ POST -> Root / "register" =>
        readJson[RegisterReq](req).flatMap {
          case Left(err) => IO.pure(err)
          case Right(RegisterReq(username, password)) =>
            register(username, password) match {
              case Left(resp) => IO.pure(resp)
              case Right(user) => Created(user.asJson)
            }
        }

      // Login
      case req @ POST -> Root / "login" =>
        readJson[LoginReq](req).flatMap {
          case Left(err) => IO.pure(err)
          case Right(LoginReq(username, password)) =>
            val user = usersByUsername.get(username)
            if (user == null || user.password != password)
              jsonError(Status.Unauthorized, "Invalid credentials")
            else {
              val token = createSession(user.id)
              val cookie = ResponseCookie(name = cookieName, content = token, path = Some("/"), httpOnly = true)
              Ok(user.asJson).map(_.addCookie(cookie))
            }
        }

      // Logout
      case req @ POST -> Root / "logout" =>
        withAuth(req) { _ =>
          req.cookies.find(_.name == cookieName).map(_.content) match {
            case Some(token) =>
              IO(invalidateSession(token)) *> Ok(Json.obj())
            case None => jsonError(Status.Unauthorized, "Authentication required")
          }
        }

      // Me
      case req @ GET -> Root / "me" =>
        withAuth(req) { user => Ok(user.asJson) }

      // Change password
      case req @ PUT -> Root / "password" =>
        withAuth(req) { user =>
          readJson[PasswordChangeReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(PasswordChangeReq(oldp, newp)) =>
              if (oldp != user.password) jsonError(Status.Unauthorized, "Invalid credentials")
              else if (newp.length < 8) jsonError(Status.BadRequest, "Password too short")
              else IO {
                val updated = user.copy(password = newp)
                usersLock.synchronized {
                  usersById.put(user.id, updated)
                  usersByUsername.put(user.username, updated)
                }
              } *> Ok(Json.obj())
          }
        }

      // List todos
      case req @ GET -> Root / "todos" =>
        withAuth(req) { user =>
          val todos = todosById.values().asScala.toVector.filter(_.ownerId == user.id).sortBy(_.id)
          Ok(Json.arr(todos.map(_.asJson): _*))
        }

      // Create todo
      case req @ POST -> Root / "todos" =>
        withAuth(req) { user =>
          readJson[CreateTodoReq](req).flatMap {
            case Left(err) => IO.pure(err)
            case Right(CreateTodoReq(title, description)) =>
              val t = title.trim
              if (t.isEmpty) jsonError(Status.BadRequest, "Title is required")
              else IO {
                val id = todoIdGen.incrementAndGet()
                val now = nowTs()
                val todo = Todo(id, t, description.getOrElse(""), completed = false, created_at = now, updated_at = now, ownerId = user.id)
                todosById.put(id, todo)
                todo
              }.flatMap(todo => Created(todo.asJson))
          }
        }

      // Get todo by id
      case req @ GET -> Root / "todos" / idStr =>
        withAuth(req) { user =>
          parseTodoId(idStr) match {
            case None => jsonError(Status.NotFound, "Todo not found")
            case Some(id) =>
              val todo = todosById.get(id)
              if (todo == null || todo.ownerId != user.id) jsonError(Status.NotFound, "Todo not found")
              else Ok(todo.asJson)
          }
        }

      // Update todo (partial)
      case req @ PUT -> Root / "todos" / idStr =>
        withAuth(req) { user =>
          parseTodoId(idStr) match {
            case None => jsonError(Status.NotFound, "Todo not found")
            case Some(id) =>
              readJson[UpdateTodoReq](req).flatMap {
                case Left(err) => IO.pure(err)
                case Right(UpdateTodoReq(titleOpt, descOpt, compOpt)) =>
                  val existing = todosById.get(id)
                  if (existing == null || existing.ownerId != user.id) jsonError(Status.NotFound, "Todo not found")
                  else {
                    titleOpt match {
                      case Some(t) if t.trim.isEmpty => jsonError(Status.BadRequest, "Title is required")
                      case _ => IO {
                        val newTitle = titleOpt.map(_.trim).filter(_.nonEmpty).getOrElse(existing.title)
                        val newDesc = descOpt.getOrElse(existing.description)
                        val newComp = compOpt.getOrElse(existing.completed)
                        val updated = existing.copy(title = newTitle, description = newDesc, completed = newComp, updated_at = nowTs())
                        todosById.put(id, updated)
                        updated
                      }.flatMap(t => Ok(t.asJson))
                    }
                  }
              }
          }
        }

      // Delete todo
      case req @ DELETE -> Root / "todos" / idStr =>
        withAuth(req) { user =>
          parseTodoId(idStr) match {
            case None => jsonError(Status.NotFound, "Todo not found")
            case Some(id) =>
              val todo = todosById.get(id)
              if (todo == null || todo.ownerId != user.id) jsonError(Status.NotFound, "Todo not found")
              else IO {
                todosById.remove(id)
              } *> NoContent()
          }
        }
    }
  }

  def run(args: List[String]): IO[ExitCode] = {
    val (port, _) = parseArgs(args)

    val httpApp = routes.orNotFound

    EmberServerBuilder.default[IO]
      .withHost(ipv4"0.0.0.0")
      .withPort(Port.fromInt(port).get)
      .withHttpApp(httpApp)
      .build
      .useForever
      .as(ExitCode.Success)
  }

  private def parseArgs(args: List[String]): (Int, List[String]) = {
    def loop(xs: List[String], port: Int, rest: List[String]): (Int, List[String]) = xs match {
      case "--port" :: p :: tail => loop(tail, p.toIntOption.getOrElse(port), rest)
      case arg :: tail            => loop(tail, port, rest :+ arg)
      case Nil                    => (if (port == 0) 8080 else port, rest)
    }
    loop(args, 0, Nil)
  }
}

File: data_mcp/gpt-5/java/rep5/src/Main.java
import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.*;
import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {
    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i = 0; i < args.length; i++) {
            if ("--port".equals(args[i]) && i + 1 < args.length) {
                port = Integer.parseInt(args[i + 1]);
                i++;
            }
        }
        InetSocketAddress addr = new InetSocketAddress("0.0.0.0", port);
        HttpServer server = HttpServer.create(addr, 0);
        TodoApp app = new TodoApp();
        server.createContext("/", app);
        ThreadPoolExecutor threadPoolExecutor = (ThreadPoolExecutor) Executors.newFixedThreadPool(16);
        server.setExecutor(threadPoolExecutor);
        server.start();
        System.out.println("Server started on 0.0.0.0:" + port);
    }
}

class TodoApp implements HttpHandler {
    private final Map<String, User> usersByUsername = new ConcurrentHashMap<>();
    private final Map<Integer, User> usersById = new ConcurrentHashMap<>();
    private final Map<String, Integer> sessions = new ConcurrentHashMap<>(); // token -> userId
    private final Map<Integer, Todo> todos = new ConcurrentHashMap<>();

    private final AtomicInteger userIdSeq = new AtomicInteger(1);
    private final AtomicInteger todoIdSeq = new AtomicInteger(1);

    private static final Pattern USERNAME_PATTERN = Pattern.compile("^[a-zA-Z0-9_]{3,50}$");

    private static final DateTimeFormatter ISO_UTC = DateTimeFormatter.ISO_INSTANT.withZone(ZoneOffset.UTC);

    @Override
    public void handle(HttpExchange exchange) throws IOException {
        try {
            route(exchange);
        } catch (Exception e) {
            // As safety net, return 500 with JSON error
            e.printStackTrace();
            sendJson(exchange, 500, Json.error("Internal server error"));
        } finally {
            try { exchange.close(); } catch (Exception ignore) {}
        }
    }

    private void route(HttpExchange exchange) throws IOException {
        String method = exchange.getRequestMethod();
        URI uri = exchange.getRequestURI();
        String path = uri.getPath();
        
        if (method.equals("POST") && path.equals("/register")) { handleRegister(exchange); return; }
        if (method.equals("POST") && path.equals("/login")) { handleLogin(exchange); return; }
        if (method.equals("POST") && path.equals("/logout")) { handleLogout(exchange); return; }
        if (method.equals("GET") && path.equals("/me")) { handleMe(exchange); return; }
        if (method.equals("PUT") && path.equals("/password")) { handlePassword(exchange); return; }
        if (path.equals("/todos")) {
            if (method.equals("GET")) { handleTodosList(exchange); return; }
            if (method.equals("POST")) { handleTodosCreate(exchange); return; }
        }
        if (path.startsWith("/todos/")) {
            String rest = path.substring("/todos/".length());
            if (rest.isEmpty() || rest.contains("/")) {
                sendJson(exchange, 404, Json.error("Not found"));
                return;
            }
            int id;
            try { id = Integer.parseInt(rest); } catch (NumberFormatException nfe) { sendJson(exchange, 404, Json.error("Not found")); return; }
            if (method.equals("GET")) { handleTodoGet(exchange, id); return; }
            if (method.equals("PUT")) { handleTodoUpdate(exchange, id); return; }
            if (method.equals("DELETE")) { handleTodoDelete(exchange, id); return; }
        }
        sendJson(exchange, 404, Json.error("Not found"));
    }

    private void handleRegister(HttpExchange exchange) throws IOException {
        String body = readBody(exchange);
        Map<String, Object> obj = Json.parseObject(body);
        String username = safeString(obj.get("username"));
        String password = safeString(obj.get("password"));

        if (username == null || !USERNAME_PATTERN.matcher(username).matches()) {
            sendJson(exchange, 400, Json.error("Invalid username"));
            return;
        }
        if (password == null || password.length() < 8) {
            sendJson(exchange, 400, Json.error("Password too short"));
            return;
        }
        synchronized (this) {
            if (usersByUsername.containsKey(username)) {
                sendJson(exchange, 409, Json.error("Username already exists"));
                return;
            }
            int id = userIdSeq.getAndIncrement();
            User user = new User(id, username, password);
            usersByUsername.put(username, user);
            usersById.put(id, user);
            sendJson(exchange, 201, user.toJsonPublic());
        }
    }

    private void handleLogin(HttpExchange exchange) throws IOException {
        String body = readBody(exchange);
        Map<String, Object> obj = Json.parseObject(body);
        String username = safeString(obj.get("username"));
        String password = safeString(obj.get("password"));
        if (username == null || password == null) {
            sendJson(exchange, 401, Json.error("Invalid credentials"));
            return;
        }
        User user = usersByUsername.get(username);
        if (user == null || !user.password.equals(password)) {
            sendJson(exchange, 401, Json.error("Invalid credentials"));
            return;
        }
        String token = UUID.randomUUID().toString().replace("-", "");
        sessions.put(token, user.id);
        Headers headers = exchange.getResponseHeaders();
        headers.add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
        sendJson(exchange, 200, user.toJsonPublic());
    }

    private void handleLogout(HttpExchange exchange) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return; // authenticate() already responded
        String token = getSessionToken(exchange);
        if (token != null) {
            sessions.remove(token);
        }
        sendJson(exchange, 200, "{}");
    }

    private void handleMe(HttpExchange exchange) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        sendJson(exchange, 200, user.toJsonPublic());
    }

    private void handlePassword(HttpExchange exchange) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        String body = readBody(exchange);
        Map<String, Object> obj = Json.parseObject(body);
        String oldp = safeString(obj.get("old_password"));
        String newp = safeString(obj.get("new_password"));
        if (oldp == null || !user.password.equals(oldp)) {
            sendJson(exchange, 401, Json.error("Invalid credentials"));
            return;
        }
        if (newp == null || newp.length() < 8) {
            sendJson(exchange, 400, Json.error("Password too short"));
            return;
        }
        user.password = newp;
        sendJson(exchange, 200, "{}");
    }

    private void handleTodosList(HttpExchange exchange) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        List<Todo> list = new ArrayList<>();
        for (Todo t : todos.values()) {
            if (t.userId == user.id) list.add(t);
        }
        list.sort(Comparator.comparingInt(t -> t.id));
        sendJson(exchange, 200, Json.todosArray(list));
    }

    private void handleTodosCreate(HttpExchange exchange) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        String body = readBody(exchange);
        Map<String, Object> obj = Json.parseObject(body);
        String title = safeString(obj.get("title"));
        String description = obj.containsKey("description") ? safeString(obj.get("description")) : "";
        if (title == null || title.trim().isEmpty()) {
            sendJson(exchange, 400, Json.error("Title is required"));
            return;
        }
        int id = todoIdSeq.getAndIncrement();
        String now = nowIso();
        Todo todo = new Todo(id, user.id, title, description == null ? "" : description, false, now, now);
        todos.put(id, todo);
        sendJson(exchange, 201, todo.toJson());
    }

    private void handleTodoGet(HttpExchange exchange, int id) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        Todo todo = todos.get(id);
        if (todo == null || todo.userId != user.id) {
            sendJson(exchange, 404, Json.error("Todo not found"));
            return;
        }
        sendJson(exchange, 200, todo.toJson());
    }

    private void handleTodoUpdate(HttpExchange exchange, int id) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        Todo todo = todos.get(id);
        if (todo == null || todo.userId != user.id) {
            sendJson(exchange, 404, Json.error("Todo not found"));
            return;
        }
        String body = readBody(exchange);
        Map<String, Object> obj = Json.parseObject(body);
        boolean modified = false;
        if (obj.containsKey("title")) {
            String title = safeString(obj.get("title"));
            if (title == null || title.trim().isEmpty()) {
                sendJson(exchange, 400, Json.error("Title is required"));
                return;
            }
            todo.title = title;
            modified = true;
        }
        if (obj.containsKey("description")) {
            String description = safeString(obj.get("description"));
            todo.description = description == null ? "" : description;
            modified = true;
        }
        if (obj.containsKey("completed")) {
            Object c = obj.get("completed");
            if (c instanceof Boolean) {
                todo.completed = (Boolean) c;
                modified = true;
            }
        }
        if (modified) {
            todo.updated_at = nowIso();
        }
        sendJson(exchange, 200, todo.toJson());
    }

    private void handleTodoDelete(HttpExchange exchange, int id) throws IOException {
        User user = authenticate(exchange);
        if (user == null) return;
        Todo todo = todos.get(id);
        if (todo == null || todo.userId != user.id) {
            // for DELETE, still return JSON error and content-type? Spec: DELETE returns no body only on success
            sendJson(exchange, 404, Json.error("Todo not found"));
            return;
        }
        todos.remove(id);
        // 204 No Content, no body
        exchange.sendResponseHeaders(204, -1);
    }

    private User authenticate(HttpExchange exchange) throws IOException {
        String token = getSessionToken(exchange);
        if (token == null) {
            sendJson(exchange, 401, Json.error("Authentication required"));
            return null;
        }
        Integer userId = sessions.get(token);
        if (userId == null) {
            sendJson(exchange, 401, Json.error("Authentication required"));
            return null;
        }
        User user = usersById.get(userId);
        if (user == null) {
            sendJson(exchange, 401, Json.error("Authentication required"));
            return null;
        }
        return user;
    }

    private String getSessionToken(HttpExchange exchange) {
        List<String> cookies = exchange.getRequestHeaders().get("Cookie");
        if (cookies == null) return null;
        for (String header : cookies) {
            String[] parts = header.split(";\\s*");
            for (String part : parts) {
                int eq = part.indexOf('=');
                if (eq > 0) {
                    String name = part.substring(0, eq).trim();
                    String value = part.substring(eq + 1).trim();
                    if (name.equals("session_id")) {
                        return value;
                    }
                }
            }
        }
        return null;
    }

    private static String nowIso() {
        Instant now = Instant.now().truncatedTo(ChronoUnit.SECONDS);
        return ISO_UTC.format(now);
    }

    private static String readBody(HttpExchange exchange) throws IOException {
        InputStream is = exchange.getRequestBody();
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buf = new byte[4096];
        int r;
        int total = 0;
        while ((r = is.read(buf)) != -1) {
            baos.write(buf, 0, r);
            total += r;
            if (total > 1024 * 1024) { // 1MB limit
                break;
            }
        }
        return baos.toString(StandardCharsets.UTF_8);
    }

    private static void sendJson(HttpExchange exchange, int status, String jsonBody) throws IOException {
        Headers headers = exchange.getResponseHeaders();
        headers.set("Content-Type", "application/json");
        byte[] bytes = jsonBody.getBytes(StandardCharsets.UTF_8);
        exchange.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    private static String safeString(Object o) {
        if (o == null) return null;
        if (o instanceof String) return (String) o;
        return String.valueOf(o);
    }
}

class User {
    public final int id;
    public final String username;
    public String password; // stored in plain for simplicity (in-memory demo)

    public User(int id, String username, String password) {
        this.id = id;
        this.username = username;
        this.password = password;
    }

    public String toJsonPublic() {
        return "{" +
                "\"id\":" + id + "," +
                "\"username\":\"" + Json.escape(username) + "\"" +
                "}";
    }
}

class Todo {
    public final int id;
    public final int userId;
    public String title;
    public String description;
    public boolean completed;
    public final String created_at;
    public String updated_at;

    public Todo(int id, int userId, String title, String description, boolean completed, String created_at, String updated_at) {
        this.id = id;
        this.userId = userId;
        this.title = title;
        this.description = description;
        this.completed = completed;
        this.created_at = created_at;
        this.updated_at = updated_at;
    }

    public String toJson() {
        StringBuilder sb = new StringBuilder();
        sb.append('{');
        sb.append("\"id\":").append(id).append(',');
        sb.append("\"title\":\"").append(Json.escape(title)).append("\",");
        sb.append("\"description\":\"").append(Json.escape(description)).append("\",");
        sb.append("\"completed\":").append(completed).append(',');
        sb.append("\"created_at\":\"").append(Json.escape(created_at)).append("\",");
        sb.append("\"updated_at\":\"").append(Json.escape(updated_at)).append("\"");
        sb.append('}');
        return sb.toString();
    }
}

class Json {
    // Very small JSON utilities for this specific API
    public static String escape(String s) {
        if (s == null) return "";
        StringBuilder sb = new StringBuilder(s.length() + 16);
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"': sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int)c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        return sb.toString();
    }

    public static String error(String msg) {
        return "{\"error\":\"" + escape(msg) + "\"}";
    }

    public static String todosArray(List<Todo> list) {
        StringBuilder sb = new StringBuilder();
        sb.append('[');
        boolean first = true;
        for (Todo t : list) {
            if (!first) sb.append(',');
            first = false;
            sb.append(t.toJson());
        }
        sb.append(']');
        return sb.toString();
    }

    public static Map<String, Object> parseObject(String s) throws IOException {
        Map<String, Object> map = new LinkedHashMap<>();
        if (s == null) return map;
        int[] idx = new int[]{0};
        skipWs(s, idx);
        if (idx[0] >= s.length() || s.charAt(idx[0]) != '{') return map; // empty map if malformed
        idx[0]++;
        skipWs(s, idx);
        while (idx[0] < s.length() && s.charAt(idx[0]) != '}') {
            skipWs(s, idx);
            String key = parseString(s, idx);
            if (key == null) { break; }
            skipWs(s, idx);
            if (idx[0] >= s.length() || s.charAt(idx[0]) != ':') break;
            idx[0]++;
            skipWs(s, idx);
            Object val = parseValue(s, idx);
            map.put(key, val);
            skipWs(s, idx);
            if (idx[0] < s.length() && s.charAt(idx[0]) == ',') {
                idx[0]++;
                skipWs(s, idx);
            } else {
                break;
            }
        }
        // move to closing brace if present
        while (idx[0] < s.length() && s.charAt(idx[0]) != '}') idx[0]++;
        if (idx[0] < s.length() && s.charAt(idx[0]) == '}') idx[0]++;
        return map;
    }

    private static void skipWs(String s, int[] idx) {
        while (idx[0] < s.length()) {
            char c = s.charAt(idx[0]);
            if (c == ' ' || c == '\n' || c == '\r' || c == '\t') idx[0]++; else break;
        }
    }

    private static String parseString(String s, int[] idx) {
        if (idx[0] >= s.length() || s.charAt(idx[0]) != '"') return null;
        idx[0]++;
        StringBuilder sb = new StringBuilder();
        while (idx[0] < s.length()) {
            char c = s.charAt(idx[0]++);
            if (c == '"') break;
            if (c == '\\') {
                if (idx[0] >= s.length()) break;
                char e = s.charAt(idx[0]++);
                switch (e) {
                    case '"': sb.append('"'); break;
                    case '\\': sb.append('\\'); break;
                    case '/': sb.append('/'); break;
                    case 'b': sb.append('\b'); break;
                    case 'f': sb.append('\f'); break;
                    case 'n': sb.append('\n'); break;
                    case 'r': sb.append('\r'); break;
                    case 't': sb.append('\t'); break;
                    case 'u':
                        if (idx[0] + 4 <= s.length()) {
                            String hex = s.substring(idx[0], idx[0] + 4);
                            try { sb.append((char) Integer.parseInt(hex, 16)); } catch (Exception ex) { /* ignore */ }
                            idx[0] += 4;
                        }
                        break;
                    default: sb.append(e); break;
                }
            } else {
                sb.append(c);
            }
        }
        return sb.toString();
    }

    private static Object parseValue(String s, int[] idx) {
        if (idx[0] >= s.length()) return null;
        char c = s.charAt(idx[0]);
        if (c == '"') return parseString(s, idx);
        if (c == 't' && s.startsWith("true", idx[0])) { idx[0] += 4; return Boolean.TRUE; }
        if (c == 'f' && s.startsWith("false", idx[0])) { idx[0] += 5; return Boolean.FALSE; }
        // null or number or other types are not expected; try to parse null
        if (c == 'n' && s.startsWith("null", idx[0])) { idx[0] += 4; return null; }
        // skip non-string primitive token
        int start = idx[0];
        while (idx[0] < s.length()) {
            char ch = s.charAt(idx[0]);
            if (ch == ',' || ch == '}' || ch == ' ' || ch == '\n' || ch == '\r' || ch == '\t' || ch == ']') break;
            idx[0]++;
        }
        String raw = s.substring(start, idx[0]).trim();
        if (raw.isEmpty()) return null;
        return raw;
    }
}
